### DISCUSSION TITLE
Processes with RA-TLS spawning child processes

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#1912] [Q&A]
Hello Gramine team, we are currently facing some questions on what would be the best approach for our use case.

We need to include RA-TLS on several components of an existing application, our initial idea was changing/forking each component of this larger application in order to include RA-TLS and then communications between these would be attested through Remote Attestation, this implies each component would live inside it's own enclave.

The problem arised when we bumped into a specific component that relies on closed source code and can't be forked in a way to include RA-TLS.
Another problem of this approach is every time there is an update to these components we need to maintain and support our RA-TLS changes on those forks.

We thought about another potential solution for our use-case but we would really appreciate some feedback on the viability and security concerns of the approach, before moving forward.

**Potential Solution would be:**

- We would build an application that just serves as proxy for each main component of the larger application
- The proxy application would be constructed in a way that is reusable for all components requiring RA-TLS
- Every component docker image would contain the executable proxy inside and the executable application service, the entrypoint would either be a shell script that spins up both proxy and the component service or the proxy itself could spawn the component service as child process. This docker image would then be graminised.
- Each component's enclave would boot from this proxy application which would spawn as child process the real component application
- RA-TLS between components would in practice be made on the proxy application itself, but since each enclave contains both the proxy and component, the attestation between components would proove that the component executable wasn't modified and it's running inside, an external party could verify this with reproducible builds.

**Our questions and concerns:**

- Does the server being hosted on child process would be visible to the host or only the main proxy process running inside the enclave?
Our concern here is that the real component shouldn't be reached from the host but instead the only way to interact with it would be through proxy component.

In case there is any other option or solution please also let us know,
Thank you!

**Comment by dimakuv** [comment]
Hello Tiago!

Yes, your solution makes sense.

One thing I'm not clear about is **when** do you want to invoke RA-TLS? Is it only at the start of the application? Or is RA-TLS supposed to be periodically invoked during the lifetime of the application?

If it's only at the start of the application, then the traditional approach (typically called "pre-main") is to:
1. Make what you call a "proxy" the entrypoint for Gramine (and thus for the Docker container).
2. The proxy invokes RA-TLS to prove it's running inside the enclave to other parties. The proxy also creates some "token" or "evidence" that it will pass to the actual application as a proof of SGX-goodness (this "token" is presented by the application to other parties when the app establishes connections with those parties). Typically the "token" is a private-public keypair (so only the application knows the private key that the proxy generated for it), or a classic X.509 certificate (again, the private key of the certificate is only known to the application).
3. At the end of this remote attestation, the proxy puts the "token"/"evidence"/"private key" in e.g. a special environment variable or as a command-line argument and invokes `execve(<actual-app>, args, envvars)` system call to start the actual application.

This is basically what a special Gramine tool does: https://gramine.readthedocs.io/en/stable/manpages/gramine-ratls.html. Maybe you can just use this tool?

> the entrypoint would either be a shell script that spins up both proxy and the component service or the proxy itself could spawn the component service as child process

Yes, both these approaches work. The shell script approach has an advantage that the proxy process doesn't "disappear" after the startup (in contrast to the approach outlined above), but has a disadvantage that you'll have three SGX enclaves now: one for the shell, one for the proxy, and one for the application. In case of `execve()`, you would only ever have one SGX enclave.

> RA-TLS between components would in practice be made on the proxy application itself, but since each enclave contains both the proxy and component, the attestation between components would proove that the component executable wasn't modified and it's running inside, an external party could verify this with reproducible builds.

Yes, correct. You're basically having an indirect chain of trust now: SGX remote attestation proves the correctness of the proxy, and by trusting the proxy code/developers, you gain trust in that it correctly starts up the desired application.

> Does the server being hosted on child process would be visible to the host or only the main proxy process running inside the enclave?

I'm not sure what you mean exactly by "visible to the host". If you mean network connections and open ports -- yes, the application (the server) will be visible to the host. Similarly, the files on which the application (the server) operates will be visible on the host.

**Reply by tiagorvmartins** [reply]
Hello @dimakuv!
Thank you for your detailed and interesting reply!

I didn't know about the gramine-ratls wrapper, it can actually help us in some specific scenarios!

> One thing I'm not clear about is when do you want to invoke RA-TLS? Is it only at the start of the application? Or is RA-TLS supposed to be periodically invoked during the lifetime of the application?

I believe the answer is during the lifetime of the application, because this application which is closed source works as a server, and it will keep receiving requests from other clients.

And because of this, we have our custom proxy which receives those requests and on every verify connection (btw verifyConnection is a hook of tlsConfig in Golang http server - this is a hook that is triggered during the handshake), we call **ra_tls_verify_callback_extended_der** using the received certificate from the peer and confirm that with contains the expected mrenclave [ _for this process we call it mutual (because we do this on both ends) authentication remote-attestation tls (maybe wrong terms/naming)_ ].

> Yes, both these approaches work. The shell script approach has an advantage that the proxy process doesn't "disappear" after the startup (in contrast to the approach outlined above), but has a disadvantage that you'll have three SGX enclaves now: one for the shell, one for the proxy, and one for the application. In case of execve(), you would only ever have one SGX enclave.

So to reiterate now, we believe we can't go down this route, because we need to verify each connection during lifetime of the application, so basically in practice we would need the three enclaves up, or maybe just two (the proxy would be entrypoint and would spawn the application).

> Yes, correct. You're basically having an indirect chain of trust now: SGX remote attestation proves the correctness of the proxy, and by trusting the proxy code/developers, you gain trust in that it correctly starts up the desired application.

That's what we thought too, nice, thank for confirming!

> I'm not sure what you mean exactly by "visible to the host". If you mean network connections and open ports -- yes, the application (the server) will be visible to the host. Similarly, the files on which the application (the server) operates will be visible on the host.

Ok, so this is where the problem exists now, if we have both servers up and running and they have those network connections visible and those ports are open to the hosts -> there is actually nothing stopping from the clients of this enclave to directly use any of the underlying closed-source-code application endpoints, instead of using the proxy that gatekeeps access, which is the one where we do the verification of the mrenclave using **ra_tls_verify_callback_extended_der**.

Not sure if I explained myself correctly or not, I hope I did, if not, please let me know and we can go deeper.

But, summarising, the keypoints we want to make from this discussion and would like guidance on possibilities:

1) We need to verify incoming connections to an existing application (so lifetime of application)
2) We can't change the existing application
3) We would like to use a proxy made by us that gatekeeps access to the existing application (using our custom verify logic relying on ra_tls_verify_callback_extended_der, that would allow only incoming connections from specific enclaves)
4) The proxy would verify those connections and on each connection if verification succeeds then would proxy that connections down to the other existing application (this could be a child process inside the same enclave)
5) The existing application endpoints **must not** be used directly, in fact those should be blocked if reached from outside the enclave, every connection needs to be validated and verified using the 'gatekeeper' proxy which runs inside the same enclave.

And basically our problem/question is on the viability of point 5, if this could ever be achieved in some way or another way different.
Again thank you so much @dimakuv and Gramine team in general for all the support!

**Reply by dimakuv** [reply]
> The existing application endpoints must not be used directly, in fact those should be blocked if reached from outside the enclave

This is an interesting problem. Currently, Gramine doesn't provide any functionality to block the application from doing such things (i.e., the network is not restricted by Gramine in any way).

What would be the solution for you? Would specifying TCP/UDP ports as explicitly blocked in the Gramine manifest make sense? Then the existing app would fail at binding to these ports. But at the same time, **all parts of your workload** (including your proxy) would also fail to bind to these ports.

I can currently only recommend to you to solve it in some software way. Maybe you could write your `LD_PRELOAD` library that will intercept some func calls from the existing application and block them?

**Reply by monavij** [reply]
Dmitrii is right that you will likely need to intercept some calls from your unmodified binary. 

Another idea we have explored in the past is around doing remote attestation enabled TLS in-line with your application. If your binary uses dynamically loaded library then maybe you can update the normal TLS library with the one that is RA enabled, so instead of doing RA-TLS first with pre-loaded lib and then TLS with the actual application you merge the two. This is assuming your app is actually using TLS to connect to clients. This is general high-level idea and actual solution will depend on your application.

Another clarification question I have is if you are doing mutual attestation? Sounds like your proxy is verifying that clients are running in an enclave and only allow connections from certain clients. Do clients also verify that your binary is running in an enclave?

**Reply by tiagorvmartins** [reply]
Thank you both for all the support, didn't want to leave this open without answering.

Anyway, to give a better answer, we did some exploration first to grasp better the solution proposed by @dimakuv and also to have a better view on what would entail.

We tried to intercept accept and accept4 functions (libc) and our custom implementations were never called, "init constructor" was being triggered (so the library was being pre loaded) but not the intercept calls of our shared object of LD_PRELOAD. 
And actually, this is not even gramine related, those calls were not being triggered even outside gramine.

Then we found that the application (that we are trying to avoid modifying) uses syscalls directly (this for launching the server - opening sockets, etc) instead of the functions accept/accept4 (from a wrapper library like libc).

And according to our searches, and please correct me if I am wrong, we can't really intercept those direct syscalls, making LD_PRELOAD option not viable for this use case.

Answering @monavij, yes we are doing mutual attestation between the several components of the stack application. But actually there is something we missing on this solution proposal, which we didn't describe:

- If we were to include a proxy application inside each component running along side with the component's unmodified app binary, there is the issue that the unmodified app is not doing attestation of the receiver (the proxy itself), so it could reply to anyone in practice if I understood correctly, making it again not viable.

We will keep exploring this, but so far seems like our last resource is really explore other applications in order to extend further and modify such that we could include our ra-tls calls on handshake.

Again, thank you for the support!

**Reply by dimakuv** [reply]
> And according to our searches, and please correct me if I am wrong, we can't really intercept those direct syscalls, making LD_PRELOAD option not viable for this use case.

This is correct, you can't intercept raw `syscall` instructions (what you called "direct syscalls").

> We will keep exploring this, but so far seems like our last resource is really explore other applications in order to extend further and modify such that we could include our ra-tls calls on handshake.

Well, you can also propose what can be (easily) modified in Gramine. If it will be sensible and simple to implement -- we'll do it.

**Reply by tiagorvmartins** [reply]
Thank you @dimakuv and Gramine team, 
That would be really helpful!! 

In case this is eventually implemented, let us know, if possible in this discussion, so we can keep track of it and apply those changes on our side and do our testing. Anyway, we will also be taking close attention to new tags here on gramine github.

In case, there are guidelines for proposal requests (github issues or somewhere) that we should follow, let us know thank you, so we add our bits to it.

Thank you!

**Reply by dimakuv** [reply]
> In case this is eventually implemented, let us know [...]

Well, my problem is that I don't know what you need from Gramine :)

**Reply by tiagorvmartins** [reply]
Sorry, missunderstood your message, 
I am not sure this is even possible from Gramine side, because it might dependant on which library the application uses for ssl, like openssl / mbedtls, etc.

But assuming it is, our feature request would be based on the following:

- unmodified application hosts an https using cert and key from path (this is big win from gramine-ratls wrapper already)
- gramine somehow would intercept all ssl attempts to handshake (would intercepting "mbedtls_ssl_handshake" from the mbedtls_gramine library even work?) then get the certificate from peer connecting, and do ra_tls_verify_callback_extended_der, using the supplied expected_mrenclave and expected_mrsigner as environment variable.

With this we would be covering the following requirements:

- We can boot any application that hosts an https server
- That server uses Gramine self-signed certificates
- That server also verifies that the incoming connections come from specific enclave
- This would be agnostic in a way that no changes are expected in the app being launched (the one that serves the https server)

In summary this would be what we need for this use case.

But I am almost certain that this is not possible to achieve, because its tied to the application itself - how that application functions on a ssl level, meaning: how the application's ssl library is used [ this would require your confirmation ]

Thank you @dimakuv

**Reply by mkow** [reply]
> I am not sure this is even possible from Gramine side, because it might dependant on which library the application uses for ssl, like openssl / mbedtls, etc.

I think what @dimakuv meant was that you can patch Gramine yourself to customize it for your specific app. I don't think it's possible / reasonable to create a flexible interface for doing that on our side which would support all the usecases.

**Reply by dimakuv** [reply]
> gramine somehow would intercept all ssl attempts to handshake (would intercepting "mbedtls_ssl_handshake" from the mbedtls_gramine library even work?) then get the certificate from peer connecting, and do ra_tls_verify_callback_extended_der, using the supplied expected_mrenclave and expected_mrsigner as environment variable.

Yes, this is not possible. SSL/TLS is above the network layer (TCP) which Gramine knows and cares about. So there's nothing Gramine can do about operations at the SSL/TLS layer.

I actually meant is that if you want e.g. a manifest option to disable certain TCP/UDP ports, this feature we could look into.
